avocado_vt/plugins/vt_joblock.py

```diff
@@ -0,0 +1,120 @@
+                                            key="show_lock_location",
+                                            key_type=bool,
+                                            default=False)
+        self.lock_file = None
+
+    def _abort(self, message, job):
+        """
+        Aborts the Job by exiting Avocado, adding a failure to the job status
+        """
+        self.log.error(message)
+        sys.exit(exit_codes.AVOCADO_JOB_FAIL | job.exitcode)
+
+    def _set_lock(self, job):
+        if not os.path.isdir(self.lock_dir):
+            msg = ('VT Job lock directory "%s" does not exist... '
+                   'exiting...' % self.lock_dir)
+            self._abort(msg, job)
+
+        pattern = 'avocado-vt-joblock-%(jobid)s-%(uid)s-%(random)s.pid'
+        # the job unique id is already random, but, let's add yet another one
+        rand = ''.join([random.choice(string.ascii_lowercase + string.digits)
+                        for i in xrange(8)])
+        path = pattern % {'jobid': job.unique_id,
+                          'uid': os.getuid(),
+                          'random': rand}
+        path = os.path.join(self.lock_dir, path)
+
+        try:
+            with open(path, 'w') as lockfile:
+                lockfile.write("%u" % os.getpid())
+            self.lock_file = path
+        except IOError as e:
+            msg = ('Failed to create VT Job lock file. Exiting...')
+            self._abort(msg, job)
+
+        if self.show_lock:
+            self.log.info("VT LOCK    : %s", self.lock_file)
+
+    def _get_lock_file(self):
+        if not os.path.isdir(self.lock_dir):
+            return None
+
+        try:
+            files = os.listdir(self.lock_dir)
+            if not files:
+                return None
+        except OSError as e:
+            if e.errno == errno.ENOENT:
+                return None
+
+        pattern = re.compile(r'avocado-vt-joblock-[0-9a-f]{40}-[0-9]+'
+                             '-[0-9a-z]{8}\.pid')
+        for lock_file in files:
+            if pattern.match(lock_file):
+                path = os.path.join(self.lock_dir, lock_file)
+                if os.path.isfile(path):
+                    content = int(open(path, 'r').read())
+                    if int(content) > 0:
+                        return path
+        return None
+
+    def _get_lock_pid(self):
+        lockfile = self._get_lock_file()
+        if lockfile is None:
+            return 0
+        return int(open(lockfile, 'r').read())
+
+    def _lock(self, job):
+        lock_pid = self._get_lock_pid()
+        if lock_pid > 0:
+            msg = ("Avocado-VT job lock acquired by PID %u. "
+                   "Aborting..." % lock_pid)
+            self._abort(msg, job)
+        self._set_lock(job)
+
+    def _unlock(self):
+        if self.lock_file:
+            os.unlink(self.lock_file)
+
+    def pre(self, job):
+        suite = job._make_test_suite(job.args.url)
```

Comment:
The whole idea of adding a custom lock to Avocado-VT as a plugin was based on the fact that Avocado itself is not yet ready to support a proper lock mechanism.  The idea of being a plugin is that it's kept further from the core code and can be disabled or replaced completely in the (hopefully) near future .

Thus, I see this as a working solution that improves the current state of things.  This is far from perfect, as I even point out its drawbacks on very clearly on the commit message.

For instance, it's *easy* to add a bunch of new hooks, as mentioned on the Job Pre/Post PR, but then this is the kind of thing we'd have to support much longer, and needs better thinking.  It's easy to propose that `pre(job)` becomes `pre(job, test_suite)` but then would it survive the test of time? Is `test_suite` really a parameter for `pre()` or should `test_suite` be available in job? This is the kind of design proposals that deserves RFC and experiments.  I would probably go with making `test_suite` a "public" job attribute, and I even coded that on Avocado.  I discarded it because it introduced bigger changes that would generate even bigger discussions.

As I said on the JobPre/Post PR, we've not promised support about at what point a job pre/post plugin is going to be called.  So we're free to tweak it for now, instead of adding one new hook for every new scenario.  Stretching it just to make a point, I don't want to have "line 201 pre job  plugin", "line 203 pre job plugin"... :smile: Since we're promising so very little interface support (just a pre/post job hook) we can and should improve the Job side of things, and then plugins would have to react to those changes (or would benefit if they react to those).

All the aspects you pointed out are indeed valid, but I had the impression that this task was supposed to implement something for Avocado-VT to fill this urgent need.  We already know that we have tasks to implement a lock API in Avocado, and to make use of it in Avocado-VT.  If I got the wrong impression and this is indeed worth a more complete re-architecture, I apologize for taking the wrong direction and suggest we close this PR completely.

@apahim and @adereis may want to chip in, since this is not so much about code itself, but about the overall project direction, and how this task fits on it.

And thanks for the review.